feat(core): add support for ambient directives

vsavkin opened the pull request:
Ambient directives can be configured when bootstraping an application.
Ambient directives can be used in every component of the application without
needing to explicitly list them.

pkozlowski-opensource commented:
Yes! I know that it can be abused by listing all the directives (thus effectively making them all "global") but we need _some_ kind of solution here, so let's try this one out.

IgorMinar commented on modules/angular2/test/public_api_spec.ts:
```diff
@@ -398,8 +399,7 @@ var NG_API = [
   'CyclicDependencyError.stackTrace',
-  'DEFAULT_PIPES',
-  'DEFAULT_PIPES_TOKEN',
+  'COMMON_PIPES',
```
what about "AMBIENT_*"?

vsavkin replied:
COMMON_ is a list of directives, and AMBIENT_ is the token.

IgorMinar replied:
right, but isn't AMBIENT_ a public api? how will ionic add their stuff to the ambient list?

IgorMinar commented on modules/angular2/test/core/linker/integration_spec.ts:
```diff
@@ -1867,8 +1869,12 @@ class PushCmpWithRef {
+  selector: 'push-cmp-with-async',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  template: '{{field | async}}',
+  pipes: [AsyncPipe]
```
why this change? isn't asyncpipe ambient?

vsavkin replied:
not in tests. I decided to remove ambient directives and pipes from tests to make them more precise.

IgorMinar replied:
good.

gkalpak commented on modules/angular2/common.ts:
```diff
@@ -0,0 +1,50 @@
+ * @Component({
+ *   selector: 'my-component',
+ *   templateUrl: 'myComponent.html',
+ *   directives: [NgClass, NgIf, NgFor, NgSwitch, NgSwitchWhen, NgSwitchDefault, NgModel, NgForm]
```
Not exactly relevant, but `OtherDirective` is missing.

vsavkin replied:
Fixed

gkalpak commented on modules/angular2/src/core/compiler/ambient.ts:
```diff
@@ -0,0 +1,57 @@
+ *
+ * ### Example
+ *
+ * Instead of writing:
```
The "instead of" section is actually missing.

gkalpak commented on modules/angular2/src/core/compiler/ambient.ts:
```diff
@@ -0,0 +1,57 @@
+ *
+ * ### Example
+ *
+ * Instead of writing:
```
Same as above (this is not the "instead of" section).

gkalpak commented on modules/angular2/src/core/compiler/ambient.ts:
````diff
@@ -0,0 +1,57 @@
+ * bootstrap(MyComponent, [provide(AMBIENT_DIRECTIVES, {useValue: [OtherDirective], multi:true})]);
+ * ```
+ */
+export const AMBIENT_DIRECTIVES = CONST_EXPR(new OpaqueToken("Ambient Directives"));
````
`OpaqueToken` type is missing.

BTW, since these are tokens, wouldn't it be more clear to call them `AMBIENT_<XYZ>_TOKEN` ? (Same as `DEFAULT_PIPES_TOKEN`.)
It doesn't read very well, but by importing `AMBIENT_PIPES`, I would expect to get some actual pipes.

vsavkin replied:
We removed the _TOKEN prefix. We can revisit it.

gkalpak commented on modules/angular2/src/core/compiler/runtime_metadata.ts:
```diff
@@ -86,18 +88,23 @@ function removeDuplicates(items: any[]): any[] {
 
-function flattenDirectives(view: ViewMetadata): Type[] {
+function flattenDirectives(view: ViewMetadata, ambientDirectives: any[]): Type[] {
   if (isBlank(view.directives)) return [];
```
Do we still want to return `[]` ? What if there are `ambintDirectives` ?

vsavkin replied:
Fixed

gkalpak commented on modules/angular2/src/core/linker/proto_view_factory.ts:
```diff
@@ -118,9 +112,13 @@ export class ProtoViewFactory {
-    _flattenList(view.pipes, pipes);
+    let pipes = [];
+    if (isPresent(this._ambientPipes)) {
+      _flattenList(this._ambientPipes, pipes);
```
Should this also be renamed to `_flattenArray` for consistency ?

vsavkin replied:
Fixed

vsavkin committed:
feat(core): add support for ambient directives

Ambient directives can be configured when bootstraping an application.
Ambient directives can be used in every component of the application without
needing to explicitly list them.

IgorMinar commented:
please resolve https://github.com/angular/angular/pull/5075#discussion_r43702776, otherwise lgtm

gkalpak commented on modules/angular2/test/public_api_spec.ts:
```diff
@@ -398,8 +400,8 @@ var NG_API = [
-  'DEFAULT_PIPES',
-  'DEFAULT_PIPES_TOKEN',
+  'COMMON_PIPES',
+  'AMBIENT_PIPES:js',
```
Probably not important, but this list used to be alphabetically sorted.

gkalpak commented:
So, if I get this correctly, `COMMON_DIRECTIVES/PIPES` will be ambient by default.

I remember this was previously decided against (for performance reasons I think).
OOC, is it that the new offline compiler will remove the perf hit (so we can use ambient directives/pipes, for improving developer ergonomics) ?

pkozlowski-opensource commented:
> I remember this was previously decided against (for performance reasons I think).

There were discussions, but perf wasn't the factor (it is compile time-thing so we are not on a critical path here).

pkozlowski-opensource commented:
The main argument against doing this was template "encapsulation": say we one day add a new thing to core directives / pipes and this might conflict with an existing directives / might start to activate on parts of HTML that were not active before.

gkalpak commented:
I see. Thx @pkozlowski-opensource !

mary-poppins commented:
Merging PR #5075 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-5075](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-5075).

vsavkin committed:
fix(playground): fix the inbox example

mary-poppins commented:
Merging PR #5075 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-5075](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-5075).
